bevy_lint/src/lints/suspicious/unit_in_bundle.rs

```diff
@@ -0,0 +1,415 @@
+//! Checks for `Bundle`s that contain the unit [`()`](unit) as a component.
+//!
+//! Specifically, this lint checks for when you pass a `Bundle` to a function or method, such as
+//! `Commands::spawn()`. If the bundle contains a unit, the lint will emit a warning.
+//!
+//! # Motivation
+//!
+//! It is possible to create bundles with a unit `()` component, since unit implements `Bundle`.
+//! Unit is not a `Component`, however, and will be ignored instead of added to the entity. Often,
+//! inserting a unit is unintentional and is a sign that the author intended to do something else.
+//!
+//! # Example
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! # use std::f32::consts::PI;
+//! #
+//! fn spawn(mut commands: Commands) {
+//!     commands.spawn(());
+//!
+//!     commands.spawn((
+//!         Name::new("Decal"),
+//!         // This is likely a mistake! `Transform::rotate_z()` returns a unit `()`, not a
+//!         // `Transform`! As such, no `Transform` will be inserted into the entity.
+//!         Transform::from_translation(Vec3::new(0.75, 0.0, 0.0))
+//!             .rotate_z(PI / 4.0),
+//!     ));
+//! }
+//! #
+//! # bevy::ecs::system::assert_is_system(spawn);
+//! ```
+//!
+//! Use instead:
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! # use std::f32::consts::PI;
+//! #
+//! fn spawn(mut commands: Commands) {
+//!     // `Commands::spawn_empty()` is preferred if you do not need to add any components.
+//!     commands.spawn_empty();
+//!
+//!     commands.spawn((
+//!         Name::new("Decal"),
+//!         // `Transform::with_rotation()` returns a `Transform`, which was the intended behavior.
+//!         Transform::from_translation(Vec3::new(0.75, 0.0, 0.0))
+//!             .with_rotation(Quat::from_rotation_z(PI / 4.0)),
+//!     ));
+//! }
+//! #
+//! # bevy::ecs::system::assert_is_system(spawn);
+//! ```
+
+use clippy_utils::{diagnostics::span_lint_hir_and_then, fn_def_id, paths::PathLookup};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, PathSegment, def_id::DefId};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, Ty};
+use rustc_span::Symbol;
+use rustc_type_ir::PredicatePolarity;
+
+use crate::{declare_bevy_lint, declare_bevy_lint_pass, paths, sym, utils::hir_parse::MethodCall};
+
+declare_bevy_lint! {
+    pub(crate) UNIT_IN_BUNDLE,
+    super::Suspicious,
+    "created a `Bundle` containing a unit `()`",
+}
+
+declare_bevy_lint_pass! {
+    pub(crate) UnitInBundle => [UNIT_IN_BUNDLE],
+}
+
+impl<'tcx> LateLintPass<'tcx> for UnitInBundle {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if expr.span.in_external_macro(cx.tcx.sess.source_map()) {
+            return;
+        }
+
+        let (fn_id, fn_args, fn_arg_types) = if let Some(MethodCall {
```

Comment:
As much as I wish, we cannot use a `match expr.kind` here because the `MethodCall` utility is a completely different type from `ExprKind::Call`. We need to use `MethodCall` here so that `can_be_spawn_empty()` doesn't miss cases like:

```rust
Commands::spawn_empty(&mut commands);
```